### DISCUSSION TITLE
Class org.mozilla.javascript.MemberBox cannot access class sun.net.www.protocol.http.HttpURLConnection

### DISCUSSION THREAD

**Discussion by Coooz** [nextgenhealthcare/connect#6273] [Q&A] [answered]
Hi,

Referring to [https://github.com/nextgenhealthcare/connect/issues/6254](url),
I have the following code that checks whether a REST API interface is available and which logs the exact same error (except my unnamed module is @30c31dd7).
```
function check_http(url) {
	//create url
	var url_check = new java.net.URL(url);
	
	//open connection to url
	var conn = url_check.openConnection();
	try{
		conn.connect();
		logger.info("HTTP Server " + url + " available.");
	}
	catch(err){
		logger.error("HTTP Server " + url + " NOT available.");
		logger.error(err.toString());
	}
	
	return conn;
}
```
The "Add a VM flag" - to mcserver.options and mcservice.options - workaround did not work, and I am not entirely sure what else to do to solve the issue. If someone can help me along here then that would be greatly appreciated. I use Mirth Connect 4.4.2. 

Thanks in advance!

**Comment by rjkroll** [comment]
Because of Java 9 and its access restrictions, and the fact Mirth uses Rhino, there are issues when trying to access an internal Java class that is not meant to be directly access by external code.  Java recommends that you try using java.net.HttpUrlConnection and its API to get around this.

EDIT:  This solution was generated by AI after I ran through the function and error:  https://www.perplexity.ai/search/here-is-my-code-in-mirth-conne-MWZBQtutR4O6K2Upzf6NfQ

You could maybe try this:

```
function check_http(url, timeout) {
    var url_check = new java.net.URL(url);
    var conn = null;
    
    try {
        conn = url_check.openConnection();
        
        if (conn instanceof java.net.HttpURLConnection) {
            conn.setConnectTimeout(timeout || 5000); // 5 seconds default timeout
            conn.connect();
            logger.info("HTTP Server " + url + " available.");
            return true;
        } else {
            logger.error("URL " + url + " is not an HTTP URL.");
            return false;
        }
    } catch (err) {
        if (err instanceof java.net.ConnectException) {
            logger.error("HTTP Server " + url + " connection refused.");
        } else if (err instanceof java.net.UnknownHostException) {
            logger.error("HTTP Server " + url + " unknown host.");
        } else {
            logger.error("HTTP Server " + url + " NOT available.");
        }
        logger.error(err.toString());
        return false;
    } finally {
        if (conn && conn instanceof java.net.HttpURLConnection) {
            conn.disconnect();
        }
    }
}
```
Along with the VMOptions possibly of :  

```
--add-opens java.base/sun.net.www.protocol.http=ALL-UNNAMED
--add-opens java.base/sun.net.www.protocol.https=ALL-UNNAMED
```

**Reply by pacmano1** [reply]
@rjkroll a few of your posts are looking a bit chatGPT/AI generated.  That's fine if course since using tools at our disposal is always smart - but maybe attribute your solution as such if that is true or perhaps post a direct link to the chatGPT discussion.    That can be helpful in the sense it shows how people work through problems using AI.

**Reply by rjkroll** [reply]
Sorry, you are correct and I do use AI to see if what possible solutions may be, and I know some may not work right away, but just trying to point people in the right direction.    In this case, I just pasted in the function and error and what I posted was what it suggested.   Would be interesting to see if it actually works, but here is the link to the conversation and I will update the original post.

https://www.perplexity.ai/search/here-is-my-code-in-mirth-conne-MWZBQtutR4O6K2Upzf6NfQ

**Reply by pacmano1** [reply]
No need to apologize.  I would say cutting and pasting into an AI tool and pasting its answer here and representing that as your own and more importantly _without testing it_ isn't really the greatest way to help folks.   If you test it and it works that is a big help to folks AI or not.

**Reply by Coooz** [reply]
@rjkroll Yes, this works absolutely fine. The vmoptions you specified are indeed necessary. Thank you!

**EDIT:**
It appeared to work with one customer. Not with other customers unfortunately.

**Comment by odoodo** [comment]
An issue was already filed for this bug [here](https://github.com/nextgenhealthcare/connect/issues/6254)

It seems to be fixed in the latest rhino version.

I worked around it by using the [Apache http client](https://hc.apache.org/httpcomponents-client-4.5.x/index.html) instead.

**Reply by Coooz** [reply]
Unfortunately, after an upgrade to version 4.4.2 with a customer of ours the bug popped up again, and the solution proposed by @rjkroll didn't work. Obviously, since the bug is fixed in the latest Rhino version (as @tonygermano states in the article linked by @odoodo), I want to make sure to have this version installed when upgrading Mirth Connect with our customers - which is something I regularly need to do.

I do not have a clue however, how to
- determine which Rhino version is installed;
- replace it with a version in which this bug is fixed, if I need to do so.

Can someone help me along here? Thanks!

**Reply by pacmano1** [reply]
Why don't you use the Apache http client?   @odoodo mentioned that already.

**Reply by Coooz** [reply]
@pacmano1 @odoodo describes using the Apache http client as a means of working around it. I prefer using a solution in which the issue is actually solved.

**Reply by Coooz** [reply]
And once more there's another customer of ours having this issue... 

I have looked at the Apache http client, but this is actually a no go. 
I must say the page to which @odoodo links is quite intimidating, I don't have _any_ experience with it, and I think it is likely the issue will appear with other customers as well which would result in some customers working with Rhino and others with Apache http client. My boss will forbid that, and rightly so in my opinion.

So, does anyone have any suggestions on how to update Rhino to a bug free version? Any help will be much appreciated!

**UPDATE ALREADY:**
Apparently the Apache http client is already built in in Mirth Connect? If so, how do I invoke it?

**Reply by pacmano1** [reply, marked as answer]
https://github.com/pacmano1/Mirth-Snippets/blob/main/loop_http_sender.js is a looping sender.  Adapt it.

**Reply by Coooz** [reply]
@pacmano1 Yes, this works fine. Thanks!

**Reply by tonygermano** [reply]
FYI, it's not possible to drop a newer Rhino into your current mirth version. Nextgen needs to update Rhino in a future mirth release.
